### PR TITLE
sql: Add attributes to role

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds role attributes to Materialize. The design of role attributes are modelled after PostgreSQL role attributes: https://www.postgresql.org/docs/15/role-attributes.html. This commit updates the CREATE ROLE statement to accept both positive and negative attributes. The ALTER ROLE statement was also added to modify role attributes.

Setting non-default role attributes are hidden behind unsafe mode, because they aren't actually checked anywhere. The attributes will be used to check privileges in a later PR.

Part of #11579

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Removes the `LOGIN` and `SUPERUSER` attributes from `CREATE ROLE`.
  - Removes the `CREATE USER` command.

### CODE REVIEW

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [inline src/adapter/src/catalog/storage.rs:395]: @mjibson Do you know if there's a better way to do this? What I'd really like to do is:
1. Get all roles out of the collection named "role" using the old role value format.
2. Delete all roles out of the collection name "role".
3. Insert some roles into the collection named "role" using the new role value format.

It seems like that should be possible but maybe we don't have the right API to do this. Instead I think I would have to spread this out over 4 releases:
release 1: Delete all values out of "roles" and insert values into "roles_new".
release 2: Delete collection "roles",
release 3: Create new collection "roles", delete all values out of "roles_new" and insert values into "roles".
release 4: Delete collection "roles_new".

**Comment by jkosh44** [comment]
This is technically blocked waiting on approval/feedback of the design doc: https://github.com/MaterializeInc/materialize/pull/17702, so feel free to ignore this until that gets merged. I just thought I'd get a head start.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:395]: I guess there's no way to do step 1 safely, because we don't know if it's a fresh installation in which case "roles" would be populated with the new value type or it's an old installation in which case "roles" would be populated with the new value type.

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [inline src/adapter/src/catalog/storage.rs:1422]: Should this function also check that the role key is of type User? Do system roles ever get removed? (If so, this probably doesn't need to be in this PR, I'm just worried about it now.)

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [inline src/adapter/src/catalog/storage.rs:1584]: Are these functions called anywhere? Here for future use?

File: src/sql/src/plan/statement/ddl.rs
Comment (mjibson) [inline src/sql/src/plan/statement/ddl.rs:2516]: Should these also just be unsupported for now? Why do they need to be unsafe?

File: src/storage-client/src/controller.rs
Comment (mjibson) [inline src/storage-client/src/controller.rs:1549]: I think this snuck in from another PR.

**Review by mjibson — COMMENTED** [review]
Looking great!

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:260]: I was wondering the same thing actually and the conclusion I came to was the same, it's useful for local dev. You bring up a good point though, if someone's username happens to be "materialize", then they'll accidentally get superuser privileges. I do think we should remove it as part of this work, but we'll have to think of a good workaround for local dev.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:376]: Unfortunately none of these migrations can ever be cleared due to how this array is indexed. There's already a comment below this saying something to this affect: 
> ... Migrations can be edited up until they ship in a
        // release, after which they must never be removed, only patched by future
       // migrations.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:395]: Will I ever be able to remove the `Option` or is that stuck there forever?

File: src/sql/src/catalog.rs
Comment (jkosh44) [reply src/sql/src/catalog.rs:296]: Actually checking any of these values is coming in a later PR, which is why I put all the non-default ones behind unsafe mode. I meant to include that in the commit message but forgot, I'll add it.

File: src/sql/src/plan/statement/ddl.rs
Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs:2516]: I used unsafe mode so that I can write tests for these, otherwise unsupported would throw errors in the tests. Though this error message can probably be improved to indicate that the role must be `LOGIN` and `SUPERUSER` and nothing else.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:1584]: `update_role` should be called for `ALTER ROLE`, but `update_roles` is never used. I'll delete `update_roles`, I just mindlessly added it.

File: src/storage-client/src/controller.rs
Comment (jkosh44) [reply src/storage-client/src/controller.rs:1549]: Unfortunately not, this function is only used for builtin migrations and silently broke at some point. Since this PR has a builtin migration it was caught.

Unfortunately, we have no good way of detecting or testing for these things without a builtin migration which is why I didn't make a separate PR.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:395]: I think I should do some serde reading. Maybe it's just smart enough to fill in default values for new bool fields. Then I can just update all of them.

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [reply src/adapter/src/catalog/storage.rs:376]: By cleared I mean making the closure empty.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs]: Removed.

File: src/sql/src/plan/statement/ddl.rs
Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs]: I've updated the error message.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:377]: Oh, my mistake. Just added that.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs]: Ok, I think I really cleaned this up using default values during deserialization. Let me know what you think.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [inline src/adapter/src/catalog/storage.rs:1967]: This struct may be pointless now, and we should instead just inline it's contents here.

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:184]: We should not merge this until we check these roles, because then we'll have roles persisted to the stash with incorrect superuser attributes.

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [inline src/adapter/src/catalog/storage.rs:388]: Why do these need to be true for the mz role? If it's a superuser then can we always make these false, including the initial migration above? I think there's a few places where the RoleAttributes struct gets made, and it might be clearer if we used a constructor with whatever params we need to do things consistently.

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [inline src/adapter/src/catalog/storage.rs:1966]: I don't understand this. What does it do?

File: src/adapter/src/catalog.rs
Comment (mjibson) [inline src/adapter/src/catalog.rs:6215]: This is a bit spooky. I'd prefer to make this struct exactly what we need going forward, and making a copy of this with an accurate data model right inside the migration closure. That can then use an `Option` instead of rust defaults, and we'll then not need to worry about remembering to remove the serde derive.

File: src/adapter/src/coord/command_handler.rs
Comment (jkosh44) [reply src/adapter/src/coord/command_handler.rs:184]: Since we have no way of knowing which roles were created with `CREATE ROLE <name> LOGIN SUPERUSER` or created by logging in via frontegg, I was planning on just grandfathering all existing roles into superuser status. Then I was going to leave it up to the organizations to revoke superuser from non-admins. Considering that, this isn't much different from the existing behavior of all new roles are given superuser status.

The other option would be to to retroactively revoke superuser status from everyone, but then there'd be no superusers left to grant it back to others.

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:388]: This is leftover from when I was going to use `materialize` as the default admin role. Though I still think this should be the case for `mz_system`. It's a bit redundant to have these be true and have `super_user` be true, but I was trying to match PostgreSQL's behavior here. They set everything to be true for the default role:
```
postgres=# SELECT * FROM pg_authid WHERE rolname = 'postgres';
 oid | rolname  | rolsuper | rolinherit | rolcreaterole | rolcreatedb | rolcanlogin | rolreplication | rolbypassrls | rolconnlimit | rolpassword | rolvaliduntil 
-----+----------+----------+------------+---------------+-------------+-------------+----------------+--------------+--------------+-------------+---------------
  10 | postgres | t        | t          | t             | t           | t           | t              | t            |           -1 |             | 
(1 row)
```

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:1966]: It essentially inline's the struct definition for serialization and deserialization. So instead of looking like this on disk:
```
{
  role: {
    name: <name>,
    attributes: { .. },
  }
}
```
It will look like this this:
```
{
  name: <name>,
  attributes: { .. },
}
```

The reason we need it, is because in all current deployments, the roles are serialized as:
```
{
  name: <name>
}
```

So we need something with `name` in the top level to deserialize old roles.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:6215]: I'm having trouble understanding this approach, can we chat about it today? As far as I know, the values are deserialized based on the structs inside the `TypedCollection`s. Is there a way to pass a new struct to a stash transaction to use for deserialization?

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs:388]: What type of params were you thinking the constructor would take? I had the same though but struggled to think of a design that worked in all contexts.

Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs]: Ok, I actually think I came up with a clean constructor approach. Let me know what you think.

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs]: Why do we need to allow users to type `CREATE ROLE <name> LOGIN` at all? They'll never be able to specify anything that passes frontegg auth. It might actually create problems because we usually only create users on first connection.

Comment (jkosh44) [reply src/adapter/src/coord/command_handler.rs]: We don't strictly need it, but this would allow organizations to set a role up with the proper permissions before that user signs in for the first time. Otherwise they can't give a user permissions until they log in for the first time.

Also, it seems like a bit of unnecessary coupling with frontegg to prevent creating login users via SQL.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs]: Ok, I updated this to reflect what we talked about to use `Option`.

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs]: What would happen if the admin granting privs forgot to add the LOGIN attribute, then that user wouldn't be able to login until they got granted that by a separate person? It's also a way of denying LOGIN to specific users that don't exist yet? Kinda weird.

**Review by mjibson — COMMENTED** [review]
New code LGTM. Should definitely hold off on merging until the design doc is in, and even then we should probably do another full overview and make sure we're not going to back ourselves into a corner. For example, we should confirm the choice to make all existing roles superuser roles, even if their frontegg role status is not currently superuser. If we do make that decision, it probably needs communication and a document for existing customers about how to un-grant superuser to those users.

I also think we need to resolve the other questions things like CREATE USER with or without LOGIN and if we even want to support one or the other, and what to do when a user shows up with frontegg auth'd creds but was previously granted a NOLOGIN role.

File: src/adapter/src/coord/command_handler.rs
Comment (jkosh44) [reply src/adapter/src/coord/command_handler.rs]: I guess this is somewhat related to the question on the design doc about syncing `SUPERUSER` status with frontegg roles. 
If a user tries to login via Frontegg, and we see that the role already exists without `LOGIN`, then do we update the role to contain `LOGIN` (i.e. sync the Materialize role with the Frontegg role)?

I think I'm actually OK with doing some Frontegg-Materialize syncing when a new connection starts up. I'm just worried about having an RBAC implementation that relies on our current frontegg usage to be correct.

**Comment by jkosh44** [comment]
@mjibson I think all open questions related to this PR have been answered and I've updated the PR accordingly. So unless I've missed something, I think this is RFAL.

**Review by mjibson — APPROVED** [review]

File: src/adapter/src/catalog/storage.rs
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs]: Nope, this is just some merge skew. I've updated it.

File: src/adapter/src/catalog/builtin.rs
Comment (jkosh44) [reply src/adapter/src/catalog/builtin.rs]: Oh yeah, that makes much more sense. Just updated.

File: src/adapter/src/catalog/builtin.rs
Comment (jkosh44) [reply src/adapter/src/catalog/builtin.rs]: That's a good idea, just updated.

File: src/sql-parser/src/parser.rs
Comment (jkosh44) [reply src/sql-parser/src/parser.rs:1638]: Just pushed an update for this.

File: src/sql/src/plan/statement/ddl.rs
Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs]: Just pushed an update for this.

**Pull request merged**
